### ISSUE TITLE
Judo screens have stopped working on iOS

### ISSUE THREAD

**Issue by SaikCaskey** [Judopay/JudoKit#179]
Hey Judo,

When I try to pop the judo screens, the view controller never appears
There was no problem until recently, and we actually didn't notice until mid-beta as we preregistered all our users and I did it using Android.

**Version/Build:**
pod 'JudoKit', '~> 6.2' (built for swift 3.2 (is there a branch for 4?))
Xcode/iOS/Swift is at latest
Swift 4 app, although I'm compiling Judo at Swift 3.2 

**How to recreate:** 
Call invokeRegisterCard
I'm testing on a 5S, but seeing the same problem on the simulator for an X etc.

**Expected behaviour:**
JudoPayViewController appears and register card flow begins

**Actual behaviour:**
Nothing happens ui-wise, and there's no hang, just the button animation I have, and then I see this get printed out every time I press my my register card button:

```
2017-12-08 11:38:15.198504+0000 ******[751:475111] === TrustKit: Successfully initialized with configuration {
    TSKPinnedDomains =     {
        "gw1.judopay.com" =         {
            TSKDisableDefaultReportUri = 0;
            TSKEnforcePinning = 1;
            TSKIncludeSubdomains = 1;
            TSKPublicKeyAlgorithms =             (
                0
            );
            TSKPublicKeyHashes = "{(\n    <******** ******** ******** ******** ******** ******** ******** ********>,\n    <******** ******** ******** ******** ******** ******** ******** ********>\n)}";
            kSKExcludeSubdomainFromParentPolicy = 0;
        };
        "judopay-sandbox.com" =         {
            TSKDisableDefaultReportUri = 0;
            TSKEnforcePinning = 1;
            TSKIncludeSubdomains = 1;
            TSKPublicKeyAlgorithms =             (
                0
            );
            TSKPublicKeyHashes = "{(\n    <******** ******** ******** ******** ******** ******** ******** ********>,\n    <******** ******** ******** ******** ******** ******** ******** ********>\n)}";
            kSKExcludeSubdomainFromParentPolicy = 0;
        };
    };
    TSKSwizzleNetworkDelegates = 0;
}
```

What gives? Have I broken something?

**Comment by zfoltin** [comment]
Hi @SaikCaskey,

Thanks for the extensive report.
Unfortunately I couldn't really reproduce the issue you are seeing. Can you share which version of the SDK are you using? Using the latest version (6.2.21 - do a `pod repo update` if pod install doesn't pull this version) the included sample app works as expected - calling `invokeRegisterCard` shows the card entry screen. Can you give the sample app a go to see if that works and what you are doing differently?

PS1: Swift 4 is coming!! Hopefully before Christmas 🎅 For now just compile the JudoKit pod with Swift 3.2
PS2: The log output you are seeing is from the certificate pinning we've included in the SDK - all looks good, you can ignore the log.

**Comment by SaikCaskey** [comment]
Hey, thanks for your response! 
I was using pod 'Judokit', '~>6.2' but I think I tried 6.2.5 at one point too, and then also the older one. I am using the latest xcode, so whichever built basically, but have now updated to 7 and I'm looking at this again. It didn't fix the issue, but perhaps I need to fiddle a bit more, I'll deintregrate my pods and then set it up again and see what happens... 
but I still think it's some breakage related to Swift. I'm debugging and seeing all your code being called, but I'm getting no ViewController appearing even though the line where it is pushed is called, and no completion handler called for invokeRegisterCard (which is what made me think it was a bug).

However, on revisiting this today, I'm seeing the screens appearing correctly in an older version of our app, the only annoying thing is that none of the judo stuff has (explicitly/directly) changed between these versions, although some swift 4 upgrading has been done, which is probably what broke it.

Given that everything works in an older version, I suspect this has nothing to do with you guys after all, and it's just some more Swift greatness (😢). 
Would you be able to offer me any tips to narrow down the problem? 
i.e., Is what I'm seeing a symptom of some mis-configuration or some other thing?

(feel free to close this when you respond unless you feel it actually is related to JudoKit)

**Comment by SaikCaskey** [comment]
Hey guys, Happy New Year!

So I'm updated to 7.0.0, haven't noticed any change, but have now isolated the issue.

It seems that when I am two levels deep in a UINavigationController hierarchy that I cannot pop the invokeRegisterCard screens. Maybe something to do with showViewController only checking the route VC? 

Something you can do to replicate my issue: 

- Add a plain VC to your sample app, set as initial, 
- embed it in a NavigationController,
- then add a button to the plain VC to present/show the original sample app content (which is already in a navigation controller). 
- Now on the main/home screen of the sample app, repeatedly tapping "Add Card" does nothing, and if I want it to work, I basically need to put the call in a dismiss handler (which does actually [seem to] work, but breaks back navigation and completion handlers obviously).

So I'm not sure - is this a swift issue, or was it just bad practice to structure UINavigationControllers like that? Apologies if I'm not Schwifty enough, I'm more experienced with Android than iOS.

Thanks for your patience!

**Comment by zfoltin** [comment]
Hey @SaikCaskey, Happy New Year to you too!

So, unfortunately the sample app is not a good example of a great Swift app - don't use it to structure your app, the main point there is to show how to usage JudoKit.

The reason the scenario you described doesn't work with the sample app is that the (Main)ViewController operates on the keyWindow.rootViewController (see `showResultsView(viewController:)` and `dismissView()` here https://github.com/Judopay/JudoKit/blob/master/Example-Swift/JudoKitSwiftExample/ViewController.swift#L374) You only want set the rootViewController when you want to reset the navigation stack. Otherwise just push VCs to the current NavigationController (or present them on top of the current VC).

Does this make sense?

**Comment by SaikCaskey** [comment]
Hey @zfoltin

Naturally, it wasn't structured around this, it was just to show you guys what I'm now seeing in my app.
It's cool if this is a different issue/not an issue, but this appears to be what's stopping my app from displaying your screens. 

Just so we're clear about this- I only call invokeRegisterCard. All the presentation logic that happens after is wholly down to you guys. 
The same thing happens in the scenario I described as now happens in my code, but it didn't seem to happen before (as now people can't register).

The only thing is that maybe having a home screen that is a navigation VC that links to another NVC that links to another NVC means I can't call Judo screens - is that to be expected? i.e., it's not fair to expect Judo to work in this scenario or scenarios like it? If so, please document things like that! 

But I'm frustrated as it worked before, flawlessly. 
I can always use your lib and make my own UI.. but I kind of really don't want to do that at this stage as I'll need to do it on Android too for consistency.

Are you saying my issue is that I have a navigation controller within a navigation controller and I shouldn't do that?

Thanks again!
Alex Caskey

**Comment by zfoltin** [comment]
Hey @SaikCaskey,

It's not very clear to me what you mean by having a navigation controller within a navigation controller. Generally you have 0 or 1 `UINavigationController` at any time on the screen. The Judo SDK checks if there is one (it is always the rootViewController if there is one) and pushes the card screen onto that, or creates a new UINavigationController on the fly and shows the card screen within that. See https://github.com/Judopay/JudoKit/blob/master/Source/JudoKit.swift#L406 for the presentation logic. Can you step through this to see where it fails in your app?

Cheers,
Zeno

**Comment by SaikCaskey** [comment]
Hey @zfoltin,

Sorry again, lack of experience conversing about swift-

I have my LandingScreenViewController (LVC) which is embedded in a NavigationViewController.
I navigate from LVC to another ViewController (VC2), which is also embedded in a NavigationViewController. It is while in the second ViewController (VC2) I want to call invokeRegisterCard.

This is basically LVC[via my SideMenu]>PaymentInfoVC(VC2)>invokeRegisterCard

What I was saying before about showViewController is that it only checks the root VC, and I was postulating that this was causing me difficulty, as when I step through the code, it goes here, calls ```navigationController.pushViewController(vc, animated: true)``` and then nothing happens after.

i.e., I have breakpoints in the JudoPayViewController viewDidLoad and these never get hit, so the UI stuff doesn't get that far.  

If I call Judo.invokeRegisterCard from LVC it works fine (but I can't navigate back naturally as it's the root). This also means I can use a dismiss completion handler in my PaymentInfo screen to clear that screen and then show Judo, but this is pointless. 

Before this started happening, it worked as expected since the day it was implemented (nearly 2 years ago!)

Let me know anywhere I should elaborate or can give examples to help you help me, 
Cheers!

**Comment by zfoltin** [comment]
Hey @SaikCaskey 

The bit that confuses me is when you say 

> I navigate from LVC to another ViewController (VC2), which is also embedded in a NavigationViewController.

Do you push VC2 into LVC's navigation controller? Or do you use a  storyboard and connect them with a segue?
It seems the right line of code gets executed (pushViewController on the navigationController) inside JudoKit. Can you check the view hierarchy (using Debug View Hierarchy) after calling `invokeRegisterCard()` to see where the card screen is?

**Comment by SaikCaskey** [comment]
Apologies @zfoltin , it's a storyboard segue! 
Also, I can't find any Judo in my view hierarchy after calling this - I imagine this is related to ```viewDidLoad()``` not even being called on the JudoPayViewController
 
So I have in my storyboard, a "Landing" VC segueing to another "Payment" VC, and there is a button here that the user taps to call ```invokeRegisterCard()```.  
(To navigate, I use a side menu/bar, I didn't mention that before, but it shouldn't matter much unless it changes something about the hierarchy, again it worked before.)

The segue presents the second VC modally, but nothing special is done 
I think I could be throwing you off by talking about VCs. It does sound like mine is actually a really basic use of the lib, there's just something slight stopping it from working.

Put simply, this is the flow:
Open App>
MainScreen>
Side Menu>
PaymentScreen>
#clicked "add card" button on payment screen>
#nothing happens, but I see logging that suggests there was no Issue>

Is that helpful at all? Perhaps you should see if you can replicate the thing I did with your sample app- as I'll bet you'll know what you're looking at better than I was

Thanks very much!
Alex Caskey

**Comment by zfoltin** [comment]
Hey @SaikCaskey 

I spent some time trying to replicate the issue in the sample app. I've even created a video to show that it works (ignore the automatically closing tableview, it is due to this line when the VC is added to a pre-existent UINavigationController: https://github.com/Judopay/JudoKit/blob/master/Example-Swift/JudoKitSwiftExample/ViewController.swift#L284):

![judo](https://user-images.githubusercontent.com/1676316/34680596-7483f5fc-f491-11e7-892c-5770e033bd50.gif)

Then I re-read your comment and spotted that you are presenting the second VC modally. I've changed the segue to be modal and now I can see the issue as you described.

Not 100% sure when it went wrong, maybe with this commit: https://github.com/Judopay/JudoKit/commit/8e318954a52799ce2c39c49249fe6c973ca662c6
Anyways, a fix will be coming soon. In the mean time, if you change the segue to `Show` instead of `Present Modally` then it should work.

Cheers for the help hunting this down,
Zeno

**Comment by zfoltin** [comment]
@SaikCaskey can you point your pod to this repo (develop branch) and see if it works? @mikehancock will release the change to cocoapods soon but I can't wait :)

**Comment by mikehancock** [comment]
I've just released 7.0.1

**Comment by SaikCaskey** [comment]
Thanks guys 👍  

I will test this out and let you know as soon as I can, although it will be tomorrow morning
I'm glad you were able to replicate it (phew) hopefully this is the end of it!

Thanks very much!
Alex Caskey

**Comment by SaikCaskey** [comment]
This fixed it guys, great work! Thanks very much @zfoltin and @mikehancock 😁 ✌️

**Comment by zfoltin** [comment]
Excellent! 👍
